project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,284 @@
+
+from pytz import timezone, utc
+
+from odoo import exceptions
+from odoo import models, fields, api, _
+from datetime import datetime
+
+
+class ProjectTask(models.Model):
+    _inherit = 'project.task'
+
+    from_days = fields.Integer(
+        string='From days',
+        help='Anticipation days from date begin or date end', default=0)
+    estimated_days = fields.Integer(
+        string='Estimated days', help='Estimated days to end', default=1,
+        oldname='anticipation_days')
+    include_in_recalculate = fields.Boolean(
+        related="stage_id.include_in_recalculate", readonly=True,
+    )
+
+    @api.constrains('estimated_days')
+    def _estimated_days_check(self):
+        for task in self:
+            if task.estimated_days <= 0:
+                raise exceptions.ValidationError(
+                    _('Estimated days must be greater than 0.')
+                )
+
+    def _dates_onchange(self, vals):
+        """
+            Try to calculate estimated_days and from_days fields
+            when date_start or date_end change.
+
+            Dates fields (date_start, date_end) have preference to
+            estimated_days and from_days
+            except if context['task_recalculate'] == True, in other words,
+            except if this change is done because task recalculating.
+        """
+        self.ensure_one()
+        # If no date changes, do nothing
+        if 'date_start' not in vals and 'date_end' not in vals:
+            return vals
+        # If we are changing dates because of task recalculating, do nothing
+        if self.env.context.get('task_recalculate'):
+            return vals
+        date_start = vals.get('date_start', self.date_start)
+        date_end = vals.get('date_end', self.date_end)
+        # If any date is False, can't calculate estimated_days nor from_days
+        if not date_start or not date_end:
+            return vals
+        from_string = fields.Datetime.from_string
+        start = from_string(date_start)
+        end = from_string(date_end)
+        resource, calendar = self._resource_calendar_select()
+        start = self._resource_timezone(start, resource)
+        end = self._resource_timezone(end, resource)
+        if end < start or not resource or not calendar:
+            return vals
+        # Calculate estimated_day
+        vals['estimated_days'] = calendar.get_working_days_of_date(
+            start_dt=start, end_dt=end, resource=resource)
+        # Calculate from_days depending on project calculation type
+        calculation_type = self.project_id.calculation_type
+        if calculation_type:
+            invert = False
+            increment = calculation_type == 'date_begin'
+            if increment:
+                if not self.project_id.date_start:
+                    # Can't calculate from_days without project date_start
+                    return vals
+                project_date = from_string(self.project_id.date_start)
```

Comment:
It should return a datetime object already, isn't it?

```suggestion
                project_date = self.project_id.date_start
```